File: cli/src/main/java/io/bitken/ss/cli/conf/ds/Manifest.java
package io.bitken.ss.cli.conf.ds;

import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.dataformat.toml.TomlMapper;
import io.bitken.ss.Build;

import java.io.IOException;
import java.nio.file.AtomicMoveNotSupportedException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.StandardCopyOption;
import java.util.Optional;

/**
 * The owned-folder marker (PB-360): {@code manifest.toml} at the data root.
 *
 * <p>Small and write-only. {@code store init} stamps it into a folder it
 * creates; {@link ProjectDataStoreResolver} reads it as a recorded fact that
 * shipsmooth owns the folder rather than inferring ownership from a {@code
 * plans/} subdirectory. Unlike {@code shipsmooth.toml} there is no upsert and
 * no unknown-but-valid content to preserve, so it is emitted from a fixed
 * template and parsed leniently — a missing or unparseable marker is simply
 * "no usable marker".
 */
public final class Manifest {

    /** The {@code kind} value marking a shipsmooth-owned state folder. */
    public static final String KIND_STATE_STORE = "state-store";
    /** The manifest's own schema version. */
    public static final String SCHEMA_VERSION = "1";

    private static final TomlMapper TOML = new TomlMapper();

    private Shipsmooth shipsmooth;
    private SchemaRef manifestSchema;

    public Shipsmooth getShipsmooth() { return shipsmooth; }
    public void setShipsmooth(Shipsmooth shipsmooth) { this.shipsmooth = shipsmooth; }

    @JsonProperty("manifest-schema")
    public SchemaRef getManifestSchema() { return manifestSchema; }
    public void setManifestSchema(SchemaRef manifestSchema) { this.manifestSchema = manifestSchema; }

    /** True when this marker names a shipsmooth-owned state folder. */
    public boolean isStateStore() {
        return shipsmooth != null && KIND_STATE_STORE.equals(shipsmooth.getKind());
    }

    /** The {@code [shipsmooth]} table. */
    public static final class Shipsmooth {
        private String kind;
        private String cliVersion;

        public String getKind() { return kind; }
        public void setKind(String kind) { this.kind = kind; }

        @JsonProperty("cli-version")
        public String getCliVersion() { return cliVersion; }
        public void setCliVersion(String cliVersion) { this.cliVersion = cliVersion; }
    }

    /** The {@code [manifest-schema]} table. */
    public static final class SchemaRef {
        private String version;

        public String getVersion() { return version; }
        public void setVersion(String version) { this.version = version; }
    }

    /**
     * Lenient read: a missing, unreadable, or unparseable file resolves as
     * {@link Optional#empty()} — the same spirit as the resolver's
     * {@code shipsmooth.toml} read path.
     */
    public static Optional<Manifest> read(Path path) {
        try {
            return Optional.ofNullable(TOML.readValue(path.toFile(), Manifest.class));
        } catch (IOException e) {
            return Optional.empty();
        }
    }

    /** The manifest body this CLI build stamps into a folder it creates. */
    public static String currentBody() {
        return render(Build.VERSION);
    }

    /** Atomically (over)write the current manifest at {@code path}. */
    public static void write(Path path) throws IOException {
        writeAtomically(path, currentBody());
    }

    /**
     * The fixed template — single-quoted literals and one blank line after
     * each table, matching the {@code shipsmooth.toml} emitter's layout so the
     * two files look alike. {@code cliVersion} is a plain semver string, never
     * containing a quote, so a literal is always safe.
     */
    private static String render(String cliVersion) {
        return "[shipsmooth]\n"
                + "kind = '" + KIND_STATE_STORE + "'\n"
                + "cli-version = '" + cliVersion + "'\n"
                + "\n"
                + "[manifest-schema]\n"
                + "version = '" + SCHEMA_VERSION + "'\n"
                + "\n";
    }

    /**
     * Sibling temp file, then atomic rename — the same guarantee
     * {@link ConfigWriter} gives {@code shipsmooth.toml} (plan-87): a failed
     * write never leaves a truncated marker behind.
     */
    private static void writeAtomically(Path target, String content) throws IOException {
        Path dir = target.toAbsolutePath().getParent();
        Files.createDirectories(dir);
        Path tmp = Files.createTempFile(dir, "manifest", ".tmp");
        try {
            Files.writeString(tmp, content);
            try {
                Files.move(tmp, target,
                        StandardCopyOption.REPLACE_EXISTING, StandardCopyOption.ATOMIC_MOVE);
            } catch (AtomicMoveNotSupportedException e) {
                Files.move(tmp, target, StandardCopyOption.REPLACE_EXISTING);
            }
        } finally {
            Files.deleteIfExists(tmp);
        }
    }
}

File: exp/rust/crates/ss-cli/src/ds/manifest.rs
//! The owned-folder marker (PB-360): `manifest.toml` at the data root.
//!
//! Small and write-only. `store init` stamps it into a folder it creates; the
//! resolver reads it as a recorded fact that shipsmooth owns the folder,
//! rather than inferring ownership from a `plans/` subdirectory. Unlike
//! `shipsmooth.toml` there is no upsert and no unknown-but-valid content to
//! preserve, so it is emitted from a fixed template and parsed leniently.

use std::path::Path;

use serde::Deserialize;

use crate::ds::atomic::write_atomically;

/// The `kind` value marking a shipsmooth-owned state folder.
pub const KIND_STATE_STORE: &str = "state-store";
/// The manifest's own schema version.
pub const SCHEMA_VERSION: &str = "1";

#[derive(Debug, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct Manifest {
    pub shipsmooth: ShipsmoothSection,
    /// Modeled (not just skipped) so `deny_unknown_fields` still accepts a
    /// real manifest; only tests read it back.
    #[serde(rename = "manifest-schema")]
    #[allow(dead_code)]
    pub manifest_schema: Option<SchemaRef>,
}

#[derive(Debug, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct ShipsmoothSection {
    pub kind: String,
    #[serde(rename = "cli-version")]
    #[allow(dead_code)]
    pub cli_version: Option<String>,
}

#[derive(Debug, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct SchemaRef {
    #[allow(dead_code)]
    pub version: Option<String>,
}

impl Manifest {
    /// True when this marker names a shipsmooth-owned state folder.
    pub fn is_state_store(&self) -> bool {
        self.shipsmooth.kind == KIND_STATE_STORE
    }

    /// Lenient read: a missing, unreadable, or unparseable file is `None` —
    /// "no usable marker" — never a hard failure (the same spirit as the
    /// resolver's `shipsmooth.toml` read).
    pub fn read(path: &Path) -> Option<Manifest> {
        let text = std::fs::read_to_string(path).ok()?;
        toml_edit::de::from_str(&text).ok()
    }

    /// The manifest body this CLI build stamps into a folder it creates.
    pub fn current_body() -> String {
        render(env!("CARGO_PKG_VERSION"))
    }

    /// Atomically (over)write the current manifest at `path`.
    pub fn write(path: &Path) -> std::io::Result<()> {
        write_atomically(path, Self::current_body().as_bytes())
    }
}

/// The fixed template — single-quoted literals and one blank line after each
/// table, matching the `shipsmooth.toml` emitter's layout so the two files
/// look alike. `cli_version` is a plain semver string, never containing a
/// quote, so a literal is always safe.
fn render(cli_version: &str) -> String {
    format!(
        "[shipsmooth]\n\
         kind = '{KIND_STATE_STORE}'\n\
         cli-version = '{cli_version}'\n\
         \n\
         [manifest-schema]\n\
         version = '{SCHEMA_VERSION}'\n\
         \n"
    )
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn current_body_has_the_expected_shape_and_the_build_version() {
        let body = Manifest::current_body();
        assert_eq!(
            body,
            format!(
                "[shipsmooth]\nkind = 'state-store'\ncli-version = '{}'\n\n\
                 [manifest-schema]\nversion = '1'\n\n",
                env!("CARGO_PKG_VERSION")
            )
        );
    }

    #[test]
    fn write_then_read_round_trips_and_classifies_as_a_state_store() {
        let tmp = tempfile::tempdir().unwrap();
        let path = tmp.path().join("sub").join("manifest.toml");

        Manifest::write(&path).unwrap();

        let read = Manifest::read(&path).expect("freshly written manifest must parse");
        assert!(read.is_state_store());
        assert_eq!(read.shipsmooth.cli_version.as_deref(), Some(env!("CARGO_PKG_VERSION")));
        assert_eq!(read.manifest_schema.unwrap().version.as_deref(), Some("1"));
    }

    #[test]
    fn write_is_idempotent_and_leaves_no_temp_litter() {
        let tmp = tempfile::tempdir().unwrap();
        let path = tmp.path().join("manifest.toml");

        Manifest::write(&path).unwrap();
        Manifest::write(&path).unwrap();

        assert_eq!(std::fs::read_to_string(&path).unwrap(), Manifest::current_body());
        let entries: Vec<_> = std::fs::read_dir(tmp.path()).unwrap().map(|e| e.unwrap().file_name()).collect();
        assert_eq!(entries, ["manifest.toml"], "no temp file left behind");
    }

    #[test]
    fn a_missing_or_unparseable_marker_reads_as_none() {
        let tmp = tempfile::tempdir().unwrap();
        let path = tmp.path().join("manifest.toml");
        assert!(Manifest::read(&path).is_none(), "missing file");

        std::fs::write(&path, "this is not toml =").unwrap();
        assert!(Manifest::read(&path).is_none(), "unparseable file");

        // A foreign but valid TOML file: no [shipsmooth] table -> None (the
        // field is required).
        std::fs::write(&path, "[other]\nx = 1\n").unwrap();
        assert!(Manifest::read(&path).is_none(), "foreign toml");
    }

    #[test]
    fn a_marker_with_an_unknown_kind_still_parses_but_is_not_a_state_store() {
        let tmp = tempfile::tempdir().unwrap();
        let path = tmp.path().join("manifest.toml");
        std::fs::write(&path, "[shipsmooth]\nkind = 'something-else'\n").unwrap();

        let read = Manifest::read(&path).expect("valid structure, unknown kind");
        assert!(!read.is_state_store());
    }
}
